types/seat/wlr_seat.c

```diff
@@ -370,3 +370,36 @@ bool wlr_seat_validate_grab_serial(struct wlr_seat *seat, uint32_t serial) {
 	//	serial == seat->touch_state.grab_serial;
 	return true;
 }
+
+void wlr_serial_add(struct wlr_serial_ringset *set, uint32_t serial) {
+	if (set->count == 0 || set->data[set->end].max_incl + 1 != serial) {
+		set->count++;
+		if (set->count > WLR_SERIAL_RINGSET_SIZE) {
+			set->count = WLR_SERIAL_RINGSET_SIZE;
+		}
+		set->end = (set->end + 1) % WLR_SERIAL_RINGSET_SIZE;
```

Comment:
Check for `set->count == 0` and don't increment, otherwise `set->data[0]` is unused on the first call.

Reply:
Instead, I initialize set->end as `WLR_SERIAL_RINGSET_SIZE - 1`, which has the same effect.

Reply:
I would prefer this check, to be honest. If these functions are reused elsewhere, that tiny initialization step is liable to be missed. Plus, setting `end = SIZE - 1` has no real semantic sense. Deferred to devs.

Reply:
I've decided to split out the case `set->count == 0`, as doing so makes the logic a bit easier to follow.